ygen/goelements_test.go

```diff
@@ -1404,3 +1410,724 @@ func TestTypeResolutionManyToOne(t *testing.T) {
 		})
 	}
 }
+
+// TestYangDefaultValueToGoDefaultValue tests the resolution of a particular
+// YANG default value to the corresponding representation in Go.
+func TestYangDefaultValueToGoDefaultValue(t *testing.T) {
+	testEnumType := yang.NewEnumType()
+	enumValues := []string{"RED", "BLUE"}
+	for _, v := range enumValues {
+		testEnumType.SetNext(v)
+		if !testEnumType.IsDefined(v) {
+			t.Fatalf("%q wasn't added to the test enum type", v)
+		}
+	}
+
+	tests := []struct {
+		name      string
+		inType    *yang.YangType
+		in        string
+		ctx       *yang.Entry
+		inEntries []*yang.Entry
+		// inEnumEntries is used to add more state for findEnumSet to test enum name generation.
+		inEnumEntries   []*yang.Entry
+		inSkipEnumDedup bool
+		inCompressPath  bool
+		want            *string
+		// wantUnionName is specified for testing the same type wrapped
+		// within a union with a de-prioritized string type.
+		wantUnionName *string
+		wantKind      yang.TypeKind
+		wantErr       bool
```

Comment:
There is a new Go style guideline on errors where it is considered brittle to compare the debug string provided by the error (I sent you the link via email). Instead, it suggests to compare against a particular property of the error (e.g. that it contains a context parameter to aid debugging), and/or create structured errors and compare using https://pkg.go.dev/github.com/google/go-cmp@v0.5.4/cmp/cmpopts#EquateErrors.

So far I agree with this, because take `yangDefaultValueToGo`, there are more than 20 errors just for a single function -- are we supposed to test them all? If the error string gets re-worded, which is sometimes unavoidable, then all affected tests have to change as well. In this very PR I've had to do this multiple times, because factoring out the restriction-checking functionality required the debug message to change.

I'm not sure the best way to handle this, I agree that from a testing perspective, it makes me rest easier that a test case failed for the same reason I wanted it to fail for, but it also increases maintenance and development burden to test the substrings. At the same time, I'm not sure how we can make error testing more robust without otherwise increasing the burden.

Reply:
I think there has definitely been changing thoughts on this. Initially (a few years ago), I think that there wasn't really a thought that there would be more structured errors. Generally, this seems a good idea such that there is a more robust means to ensure that we're capturing the right error case (which I would argue helps with understanding that we're testing) without being tied to a particular string.

Let's stick with `wantErr` (binary error check here) today, and then we can discuss whether we want to try and move towards something like a more structured error going forward.